src/autopas/options/Option.h

```diff
@@ -0,0 +1,133 @@
+/**
+ * @file Option.h
+ * @author F. Gratl
+ * @date 10/14/19
+ */
+
+#pragma once
+
+#include <map>
+#include <set>
+#include <string>
+
+#include "autopas/utils/StringUtils.h"
+
+namespace autopas {
+
+/**
+ * Base class for autopas options.
+ * @tparam actualOption Curiously recurring template pattern.
+ */
+template <typename actualOption>
+class Option {
+ public:
+  /**
+   * Constructor.
+   */
+  constexpr Option() = default;
+
+  /**
+   * No cast to bool.
+   * @return
+   */
+  explicit operator bool() = delete;
+
+  /**
+   * Provides a way to iterate over the possible options.
+   * @return set of all possible values of this option type.
+   */
+  static std::set<actualOption> getAllOptions() {
+    std::set<actualOption> retSet;
+    auto mapOptionNames = actualOption::getOptionNames();
+    std::for_each(mapOptionNames.begin(), mapOptionNames.end(),
+                  [&retSet](auto pairOpStr) { retSet.insert(pairOpStr.first); });
+    return retSet;
+  };
+
+  /**
+   * Converts an Option object to its respective string representation.
+   * @return The string representation or "Unknown Option (<IntValue>)".
+   */
+  std::string to_string() const {
+    auto &actualThis = *static_cast<const actualOption *>(this);
+    auto mapOptNames = actualOption::getOptionNames();  // <- not copying the map destroys the strings
+    auto match = mapOptNames.find(actualThis);
+    if (match == mapOptNames.end()) {
+      return "Unknown Option (" + std::to_string(actualThis) + ")";
+    } else {
+      return match->second;
+    }
+  }
+
+  /**
+   * Converts a string of options to a set of enums. For best results, the options are expected to be lower case.
```

Comment:
```
For best results, the options are expected to be lower case.
```
As far as I see this, a better description would be sth. like:
```
The algorithm is case-insensitive.
```
Or sth. like this, as you are anyways transforming all options to lowercase.

Reply:
The algorithm actually is case sensitive. However, since the idea for the new parsing process is that the user does not need to know exactly how anything is spelled we compare only to lowercase. The argument `optionsString` however, is not transformed to lowercase since this would be redundant most of the time. Therefore the algorithm will probably work when given a mixed-case string but not as good as when presented with a lower case one.
Example: when given `AoS` it is unclear if `soa` or `aos` is meant.

Reply:
oh right, this looked different to me, as you were doing that lowercase transformation twice. so i thought you did this also for the input...